src/registry/mod.rs

```diff
@@ -1,61 +1,210 @@
 mod asynchronous;
 mod synchronous;
 
+use std::{
+    cell::{Ref, RefCell},
+    iter::Rev,
+    ops::Range,
+    rc::Rc,
+};
+
+use crate::source::FileSource;
+
+use chunky_vec::ChunkyVec;
+use fluent::FluentResource;
+use fluent_fallback::BundleGenerator;
 use itertools::Itertools;
 use unic_langid::LanguageIdentifier;
 
-use crate::source::FileSource;
+pub use asynchronous::GenerateBundles;
+pub use synchronous::GenerateBundlesSync;
+
+pub type FluentResourceSet = Vec<Rc<FluentResource>>;
+
+/// Generate a permutation of all registered source file indices for `length`
+/// in reverse order. ie. The last source added to the registry with `add_source`
+/// is returned first.
+pub(crate) fn permute_iter(
+    source_count: usize,
+    length: usize,
+) -> itertools::MultiProduct<Rev<Range<usize>>> {
+    (0..length)
+        .map(|_| (0..source_count).rev())
+        .multi_cartesian_product()
+}
 
 #[derive(Default)]
+struct Shared {
+    lang_ids: Vec<LanguageIdentifier>,
+    sources: RefCell<ChunkyVec<FileSource>>,
+}
+
+pub struct L10nRegistryLocked<'a> {
+    lock: Ref<'a, ChunkyVec<FileSource>>,
+}
+
+impl<'a> L10nRegistryLocked<'a> {
+    pub fn iter(&self) -> impl Iterator<Item = &FileSource> {
+        self.lock.iter()
+    }
+
+    pub fn len(&self) -> usize {
+        self.lock.len()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    pub fn source_idx(&self, index: usize) -> &FileSource {
+        self.lock.get(index).expect("Index out-of-range")
+    }
+
+    pub fn get_source(&self, name: &str) -> Option<&FileSource> {
+        self.lock.iter().find(|&source| source.name == name)
+    }
+
+    pub fn generate_sources_for_file<'l>(
+        &'l self,
+        langid: &'l LanguageIdentifier,
+        res_id: &'l str,
+    ) -> impl Iterator<Item = &FileSource> {
```

Comment:
do you need the type annotations if all uses the same?

Reply:
With out the `'l` I get compilation errors:

```
lifetime mismatch
...but data from `langid` is returned hererustc(E0623)
mod.rs(69, 17): this parameter and the return type are declared with different lifetimes...
mod.rs(71, 10): 
mod.rs(71, 10): Error originated from macro here
```